packages/truffle-codec-utils/src/types/abify.ts

```diff
@@ -0,0 +1,242 @@
+import debugModule from "debug";
+const debug = debugModule("codec-utils:types:abify");
+
+import { Types } from "./types";
+import { Values } from "./values";
+import { Errors } from "./errors";
+import { UnknownUserDefinedTypeError } from "../errors";
+import BN from "bn.js";
+import { Conversion as ConversionUtils } from "../conversion";
+
+export function abifyType(dataType: Types.Type, userDefinedTypes?: Types.TypesById): Types.Type | undefined {
+  switch(dataType.typeClass) {
+    //we only need to specially handle types that don't go in
+    //the ABI, or that have some information loss when going
+    //in the ABI
+    //note that we do need to handle arrays, due to recursion!
+    //First: types that do not go in the ABI
+    case "mapping":
+    case "magic":
+      return undefined;
+    //Next: address & contract, these can get handled together
+    case "address":
+    case "contract":
+      return {
+        typeClass: "address",
+        payable: false,
```

Comment:
false in all cases? really?

Reply:
Because we're converting from the ABI, and there's no such ABI type as `address payable`, only `address`.

Now you'll recall I did discuss with you a while back having address-of-unspecified-payability as being a thing in the type system in order to make ABI-mode and full-mode more compatible, but, well, ultimately that compatibility turned out to be a fool's errand and I didn't do it, so I didn't make address-of-unspecified-payability a thing either.  I admit it's maybe a little misleading, but, well, that's ABI types for you.

I mean, alternatively you could make it always `true`, but, that seems even worse.  I think it's fine as-is.

Reply:
Always being true seems more correct to me? Since the ABI treats everything as payable.

Reply:
I mean, maybe?  IMO it'd be a little more confusing to have `address payable` turn up as a type in your ABI-mode decoding.  OTOH we did do something similar regarding plain uses of `address` in pre-0.5.0 Solidity.  I'm more inclined towards `false`, but if you really think `true` is the way to go we can change it.  Just hoping nobody ends up relying on the result. :P

Reply:
Alternatively, could we make `payable` nullable? Like... we don't really know!

Reply:
Hahaha yup that possibility again!  I mean, we could?  This is basically the possibility I discussed with you before.  I'm a little leery of that, as it makes `address` (the Solidity type) and `address` (the ABI type) into different things?  (Also, we might want to add a tag to distinguish the two cases, which was actually how we discussed doing it before.)  But yeah that's definitely one more possibility?

Reply:
Hm, but aren't they different things?

I'm not sure how I feel about this. I see the trickiness. We might just have to solve this with notes in documentation that this isn't supported yet.

Reply:
I mean, they're *kind* of different, but I didn't want to get into, like, distinguishing ABI types from Solidity types except where necessary.  Like, is a `uint8` different when it's a Solidity `uint8` vs an ABI `uint8`?  (Because the latter might secretly be an enum.)  I thought it was best to say, no it's not, and have the `typeHint` field for when we do have that information.

Not sure what you mean by "not supported yet".  It will very definitely be supported, the only question is which possibility we go with!

Actually I only just noticed that this is in `abify.ts`, that this is `abifyType`, and not `abiParameterToType` like I thought.  So, in this context, if you really want that information as to whether it was payable or not, you can check the `typeHint` field!

Reply:
Actually, that's another slight difference between using `abify` vs actually running the decoder in ABI mode: If you use `abify`, you will get extra type information in `typeHint`, like whether the address was payable or not.  If you actually run the decoder in ABI mode, you won't.

Reply:
OK, after talking to @gnidan we decided to make `payable` nullable, marking it with `kind: "general"` or `kind: "specific"`, like with external functions.